File: bot-api/src/main/java/gg/mineral/bot/api/math/Positionable.java
package gg.mineral.bot.api.math;

import lombok.val;
import org.eclipse.jdt.annotation.NonNull;

public interface Positionable {

    /**
     * Gets the X position.
     *
     * @return the X position
     */
    double getX();

    /**
     * Gets the Y position.
     *
     * @return the Y position
     */
    double getY();

    /**
     * Gets the Z position.
     *
     * @return the Z position
     */
    double getZ();

    default double distance3DTo(@NonNull Positionable entityIn) {
        return Math.sqrt(distance3DToSq(entityIn));
    }

    default double distance3DToSq(@NonNull Positionable entityIn) {
        val dX = this.getX() - entityIn.getX();
        val dY = this.getY() - entityIn.getY();
        val dZ = this.getZ() - entityIn.getZ();
        return dX * dX + dY * dY + dZ * dZ;
    }

    default double distance3DToSq(double x, double y, double z) {
        val dX = this.getX() - x;
        val dY = this.getY() - y;
        val dZ = this.getZ() - z;
        return dX * dX + dY * dY + dZ * dZ;
    }

    default double distance2DToSq(double x, double z) {
        val dX = this.getX() - x;
        val dZ = this.getZ() - z;
        return dX * dX + dZ * dZ;
    }

    default double distance2DTo(double x, double z) {
        return Math.sqrt(distance2DToSq(x, z));
    }
}

File: bot-api/src/main/kotlin/gg/mineral/bot/api/math/Positionable.kt
package gg.mineral.bot.api.math

import kotlin.math.sqrt

@JvmDefaultWithCompatibility
interface Positionable {
    /**
     * Gets the X position.
     *
     * @return the X position
     */
    val x: Double

    /**
     * Gets the Y position.
     *
     * @return the Y position
     */
    val y: Double

    /**
     * Gets the Z position.
     *
     * @return the Z position
     */
    val z: Double

    fun distance3DTo(entityIn: Positionable): Double {
        return sqrt(distance3DToSq(entityIn))
    }
    
    fun distance3DToSq(entityIn: Positionable): Double {
        val dX = this.x - entityIn.x
        val dY = this.y - entityIn.y
        val dZ = this.z - entityIn.z
        return dX * dX + dY * dY + dZ * dZ
    }

    fun distance3DToSq(x: Double, y: Double, z: Double): Double {
        val dX = this.x - x
        val dY = this.y - y
        val dZ = this.z - z
        return dX * dX + dY * dY + dZ * dZ
    }

    fun distance2DToSq(x: Double, z: Double): Double {
        val dX = this.x - x
        val dZ = this.z - z
        return dX * dX + dZ * dZ
    }

    fun distance2DTo(x: Double, z: Double): Double {
        return sqrt(distance2DToSq(x, z))
    }
}
